explainer/index.bs

```diff
@@ -26,10 +26,139 @@ See [Introduction](https://gpuweb.github.io/gpuweb/#introduction).
+Most of WebGPU's validation rules are necessary to ensure it is secure to use, so all the validation needs to happen in the GPU process.
+
+Likewise, all GPU driver objects only live in the GPU process, including large allocations (like buffers and textures) and complex objects (like pipelines).
+In the content process, WebGPU types (`GPUBuffer`, `GPUTexture`, `GPURenderPipeline`, ...) are mostly just "handles" that identify objects that live in the GPU process.
+This means that the CPU and GPU memory used by WebGPU object isn't necessarily known in the content process.
+A `GPUBuffer` object can use maybe 150 bytes of CPU memory in the content process but hold a 1GB allocation of GPU memory.
+
+See also the description of [the content and device timelines in the specification](https://gpuweb.github.io/gpuweb/#programming-model-timelines).
+
 # JavaScript API # {#api}
 
 ## Bitflags ## {#bitflags}
 
+## Object Validity and Destroyed-ness ## {#invalid-and-destroyed}
+
+### WebGPU's Error Monad #### {#error-monad}
+
+A.k.a. Contagious Internal Nullability.
+A.k.a. transparent [promise pipelining](http://erights.org/elib/distrib/pipeline.html).
+
+WebGPU is a very chatty API, with some applications making tens of thousands of calls per frame to render complex scenes.
+We have seen that the GPU processes needs to validate the commands to satisfy their security property.
+To avoid the overhead of validating commands twice in both the GPU and content process, WebGPU is designed so Javascript calls can be forwarded directly to the GPU process and validated there.
+See the error section for more details on what's validated where and how errors are reported.
+
+At the same time, during a single frame WebGPU objects can be created that depend on one another.
+For example a `GPUCommandBuffer` can be recorded with commands that use temporary `GPUBuffer`s created in the same frame.
+In this example, because of the performance constraint of WebGPU, it is not possible to send the message to create the `GPUBuffer` to the GPU process and synchronously wait for its processing before continuing Javascript execution.
+
+Instead, in WebGPU all objects (like `GPUBuffer`) are created immediately on the content timeline and returned to JavaScript.
+The validation is almost all done asynchronously on the "device timeline".
+In the good case, when no errors occur (validation or out-of-memory), everything looks to JS as if it is synchronous.
+However, when an error occurs, the object is tagged as invalid on the GPU process side.
+All WebGPU calls validate that all their arguments are valid objects.
+Calls become no-ops (aside from error reporting) when validation fails.
+If the call returns a WebGPU object, then the returned object is also invalid (hence the term "contagious").
+
+<img src="https://i.imgur.com/DY9GheM.png"/>
```

Comment:
Did you happen to draw this in something that can export an SVG? Possibly smaller than a PNG and text could be searchable/selectable (I think) and editable

Reply:
Yeah I'll export an SVG and make a PR with it.